### ISSUE TITLE
[WebProfilerBundle] `/_wdt/styles.css` is not found

### ISSUE THREAD

**Issue by alexislefebvre** [symfony/symfony#59045]
### Symfony version(s) affected

7.2.0

### Description

When I browse my local website, the debug toolbar is displayed without any style:

![image](https://github.com/user-attachments/assets/4e6135d3-0585-4bba-a223-0acaeba4c87f)

Loading `http://localhost:8000/_wdt/styles.css` in the browser shows:

> The requested resource /_wdt/styles.css was not found on this server.

For some reason, it is passed to the PHP built-in server:

![image](https://github.com/user-attachments/assets/85022a46-8b8c-4cb0-aa16-b707665d67ce)

### How to reproduce

The PHP server runs in Docker through `php -S 0.0.0.0:8000 -t public/`.

It looks like it's not the source of the problem since `http://localhost:8000/_wdt/50aa18?XDEBUG_IGNORE=1` works without any issue.

### Possible Solution

See this workaround to change the path of the path in your project: https://github.com/symfony/symfony/issues/59045#issuecomment-2523839740

### Additional Context

The debug toolbar has no issues:

![image](https://github.com/user-attachments/assets/7c3630d0-dabf-45d4-9843-e816bb912254)

#### Router

`config/routes/web_profiler.yaml`

```yaml
when@dev:
    web_profiler_wdt:
        resource: '@WebProfilerBundle/Resources/config/routing/wdt.xml'
        prefix: /_wdt

    web_profiler_profiler:
        resource: '@WebProfilerBundle/Resources/config/routing/profiler.xml'
        prefix: /_profiler
```

`debug:router`:

```
$ bin/console debug:router _wdt_stylesheet
+--------------+--------------------------------------------------------------------------+
| Property     | Value                                                                    |
+--------------+--------------------------------------------------------------------------+
| Route Name   | _wdt_stylesheet                                                          |
| Path         | /_wdt/styles.css                                                         |
| Path Regex   | {^/_wdt/styles\.css$}sDu                                                 |
| Host         | ANY                                                                      |
| Host Regex   |                                                                          |
| Scheme       | ANY                                                                      |
| Method       | ANY                                                                      |
| Requirements | NO CUSTOM                                                                |
| Class        | Symfony\Component\Routing\Route                                          |
| Defaults     | _controller: web_profiler.controller.profiler::toolbarStylesheetAction() |
| Options      | compiler_class: Symfony\Component\Routing\RouteCompiler                  |
|              | utf8: true                                                               |
+--------------+--------------------------------------------------------------------------+

$ bin/console debug:router _wdt
+--------------+----------------------------------------------------------------+
| Property     | Value                                                          |
+--------------+----------------------------------------------------------------+
| Route Name   | _wdt                                                           |
| Path         | /_wdt/{token}                                                  |
| Path Regex   | {^/_wdt/(?P<token>[^/]++)$}sDu                                 |
| Host         | ANY                                                            |
| Host Regex   |                                                                |
| Scheme       | ANY                                                            |
| Method       | ANY                                                            |
| Requirements | NO CUSTOM                                                      |
| Class        | Symfony\Component\Routing\Route                                |
| Defaults     | _controller: web_profiler.controller.profiler::toolbarAction() |
| Options      | compiler_class: Symfony\Component\Routing\RouteCompiler        |
|              | utf8: true                                                     |
+--------------+----------------------------------------------------------------+
```

Related:

- #58287

**Comment by alexisLefebvre** [comment]
On this line: https://github.com/symfony/symfony/blob/4ed9d032c7b61ae49d526565bbc8fdbcf76d74c3/src/Symfony/Bundle/WebProfilerBundle/Resources/config/routing/wdt.xml#L7

Replacing `/styles.css` with `/styles` fixes the issue: the CSS is rendered when calling `/_wdt/styles`.

So it looks like it's a limitation of the built-in server of PHP. My hypothesis: it tries to find the file `/_wdt/styles.css` when this path is provided, but call Symfony when the path is changed to `/_wdt/styles`.

**Comment by alexisLefebvre** [comment]
Possible (ugly) workaround: add this in your template:

```
        {% if 'dev' == app.request.server.get('APP_ENV') %}
            <style>
                {{ include('@WebProfiler/Profiler/toolbar.css.twig') }}
            </style>
        {% endif %}
```

**Comment by alexisLefebvre** [comment]
It looks like it's an edge case, let's see if someone else has the same issue.

**Comment by lugosium** [comment]
Got this issue too

**Comment by alexisLefebvre** [comment]
> Got this issue too

Here is a workaround by using the [old version](https://github.com/symfony/symfony/blob/31b6a568850ad260b9d20122667a84ce0665ba7a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig) to override the new version:

1. create the `templates/bundles/WebProfilerBundle/Profiler/` directory
2. copy-paste the old version of the Twig file in the `toolbar_js.html.twig` file in this new directory

Through CLI:

```bash
mkdir --parents templates/bundles/WebProfilerBundle/Profiler/
curl https://raw.githubusercontent.com/symfony/symfony/31b6a568850ad260b9d20122667a84ce0665ba7a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig \
--output templates/bundles/WebProfilerBundle/Profiler/toolbar_js.html.twig
```

This will revert [this change](https://github.com/symfony/symfony/pull/58287/commits/c36fcff655fc31c65bacc7d973e1f61936cc93c9#diff-4ea86c834e4836a0ea7983c47068529a10ae4b54620e0cb094b53b4644c418b1L12) and inline the CSS instead of loading it through an URL.

**Comment by alexisLefebvre** [comment]
It would be nice to be able to use the "old" way with a parameter of WebProfilerBundle.

**Comment by alexisLefebvre** [comment]
> Got this issue too

What server do you use? The built-in PHP server too?

**Comment by alexisLefebvre** [comment]
Another workaround: use Symfony CLI to run the server: https://symfony.com/download

In the context of PHP in a Docker container, based on an Alpine image:

1. Install it:
    ```diff
    + RUN apk add --no-cache bash
    + RUN wget https://get.symfony.com/cli/installer -O - | bash
    + RUN mv -v /root/.symfony5/bin/symfony /usr/local/bin/symfony
    ```
2. Start the server by replacing the `entrypoint` of Docker:
    ```diff
    - entrypoint: php -S 0.0.0.0:8000 -t public/
    + entrypoint: symfony local:server:start --port=8000 --allow-all-ip --no-tls
    ```
3. Build and restart the Docker container
4. It works!

**Comment by lugosium** [comment]
> > Got this issue too
> 
> What server do you use? The built-in PHP server too?

I'm behind docker/nginx.

Thanks for your help 👍

**Comment by lugosium** [comment]
Issue is (for me) in the nginx vhost:

```
    location ~* \.(jpg|jpeg|png|webp|gif|ico|css|js|woff|woff2|ttf)$ {
       expires 1d;
    }
```

If I remove css in the regex it's ok. 
But i don't really want to remove that...

**Comment by alexisLefebvre** [comment]
### Actual solution with PHP's built-in server:

```diff
- php -S 0.0.0.0:8000 -t public/
+ php -S 0.0.0.0:8000 public/index.php
```

~~The last parameter is the router, it will serve the file through Symfony.~~

`-t public/` (the docroot) has to be removed, `public/index.php` will be used as a router, and the CSS file for WebProfilerBundle will be served by Symfony.

**Comment by rpkamp** [comment]
Same issue here, behind NGiNX + PHP-FPM. I get a 404 on `/_wdt/styles.css`.

Probably caused by 
```
location ~* \.(?:css|js|jpg|jpeg|gif|png|ico|svg|ttf|otf|woff|woff2|eot)$ {
    expires 1y;
    add_header Cache-Control "public, immutable";
}
```

telling NGiNX all `css` files should be served by it, and it will not try PHP to serve it.

**Comment by alexisLefebvre** [comment]
It works as expected now.

@lugosium I didn’t get if you use PHP’s built-in server or something else (like `php-fpm` or `apache`).

If you use the built-in server, please try the solution on the comment above.

Otherwise, please open another issue since it may be related to nginx.

**Comment by rpkamp** [comment]
Adding this to my NGiNX config seems to solve it, though I don't really like having to configure something explicitly for a dev tool.

```
location ^~ /_wdt {
    try_files $uri /index.php?$query_string;
}
```

**Comment by rpkamp** [comment]
Chaning `/styles.css` to `/styles` would also fix the issue, as suggested [here](https://github.com/symfony/symfony/issues/59045#issuecomment-2508111347)

**Comment by lugosium** [comment]
> Chaning `/styles.css` to `/styles` would also fix the issue, as suggested [here](https://github.com/symfony/symfony/issues/59045#issuecomment-2508111347)

+1 here, suffix `.css` not needed as we can rely on the `'Content-Type' => 'text/css'` header and it's not a file anymore.

**Comment by alexisLefebvre** [comment]
I reopen this issue so that Symfony's maintainers can study this case.

**Comment by andrewbrereton** [comment]
> Adding this to my NGiNX config seems to solve it, though I don't really like having to configure something explicitly for a dev tool.
> 
> ```
> location ^~ /_wdt {
>     try_files $uri /index.php?$query_string;
> }
> ```

This worked for me.

Running nginx php-fpm inside Docker.

**Comment by Jean-Gian** [comment]
I got this issue too, with apache

**Comment by PhilETaylor** [comment]
Are you all "behind" a proxy and not set `trusted_proxies`? I think I had this issue yesterday

**Comment by lugosium** [comment]
No proxy for me.

**Comment by mustanggb** [comment]
I encountered this upgrading from 7.1.9 to 7.2.0, the fix from [rpkamp](https://github.com/symfony/symfony/issues/59045#issuecomment-2511755559) worked nicely.

**Comment by alexisLefebvre** [comment]
Another workaround:

> On this line:
> 
> https://github.com/symfony/symfony/blob/4ed9d032c7b61ae49d526565bbc8fdbcf76d74c3/src/Symfony/Bundle/WebProfilerBundle/Resources/config/routing/wdt.xml#L7
> 
> Replacing `/styles.css` with `/styles` fixes the issue: the CSS is rendered when calling `/_wdt/styles`.

I was able to do this in a project, by changing the config instead of the file in `vendor/`:

`config/routes/web_profiler.yaml`

```diff
    when@dev:
        web_profiler_wdt:
            resource: '@WebProfilerBundle/Resources/config/routing/wdt.xml'
            prefix: /_wdt
    
        web_profiler_profiler:
            resource: '@WebProfilerBundle/Resources/config/routing/profiler.xml'
            prefix: /_profiler
    
+       _wdt_stylesheet:
+           path: /custom_wdt/styles
+           controller: web_profiler.controller.profiler::toolbarStylesheetAction
```

It may require to delete the cache.

**Comment by THenkeDE** [comment]
same problem with local WAMP Environment on Win11 with standard Apache (no nGinx).

Test-Updated from working 7.1.* to 7.3.*, nothing else changed.

**Comment by Power2All** [comment]
I had the same issue when updating 7.1 to 7.2
https://github.com/symfony/symfony/issues/59045#issuecomment-2508111347 fixed the issue after clearing the cache, development running purely in PHP 8.3 (using built-in web server), and with Apache as well. Should also work fine in NGINX.

**Comment by Jean-Gian** [comment]
Solution by @alexislefebvre works perfectly for me

> Another workaround:
> 
> > On this line:
> > https://github.com/symfony/symfony/blob/4ed9d032c7b61ae49d526565bbc8fdbcf76d74c3/src/Symfony/Bundle/WebProfilerBundle/Resources/config/routing/wdt.xml#L7
> > 
> > Replacing `/styles.css` with `/styles` fixes the issue: the CSS is rendered when calling `/_wdt/styles`.
> 
> I was able to do this in a project:
> 
> `config/routes/web_profiler.yaml`
> 
> ```diff
>     when@dev:
>         web_profiler_wdt:
>             resource: '@WebProfilerBundle/Resources/config/routing/wdt.xml'
>             prefix: /_wdt
>     
>         web_profiler_profiler:
>             resource: '@WebProfilerBundle/Resources/config/routing/profiler.xml'
>             prefix: /_profiler
>     
> +       _wdt_stylesheet:
> +           path: /custom_wdt/styles
> +           controller: web_profiler.controller.profiler::toolbarStylesheetAction
> ```
> 
> It may require to delete the cache.

**Comment by Power2All** [comment]
> Solution by @alexislefebvre works perfectly for me
> 
> > Another workaround:
> > > On this line:
> > > https://github.com/symfony/symfony/blob/4ed9d032c7b61ae49d526565bbc8fdbcf76d74c3/src/Symfony/Bundle/WebProfilerBundle/Resources/config/routing/wdt.xml#L7
> > > 
> > > Replacing `/styles.css` with `/styles` fixes the issue: the CSS is rendered when calling `/_wdt/styles`.
> > 
> > 
> > I was able to do this in a project:
> > `config/routes/web_profiler.yaml`
> > ```diff
> >     when@dev:
> >         web_profiler_wdt:
> >             resource: '@WebProfilerBundle/Resources/config/routing/wdt.xml'
> >             prefix: /_wdt
> >     
> >         web_profiler_profiler:
> >             resource: '@WebProfilerBundle/Resources/config/routing/profiler.xml'
> >             prefix: /_profiler
> >     
> > +       _wdt_stylesheet:
> > +           path: /custom_wdt/styles
> > +           controller: web_profiler.controller.profiler::toolbarStylesheetAction
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It may require to delete the cache.

I didn't need to do any of that, only the wdt.xml file was enough, and clearing cache.
Can you try it without that part ?

**Comment by alexisLefebvre** [comment]
> I didn't need to do any of that, only the wdt.xml file was enough, and clearing cache.
> Can you try it without that part ?

The `wdt.xml` is in the `vendor/` directory so it only works on one install and can't be deployed as is.

**Comment by Jean-Gian** [comment]
The solution that @alexislefebvre suggested is stable and clean, is agnostic to the webserver and doesn't rely on editing vendor files

**Comment by Power2All** [comment]
> The solution that @alexislefebvre suggested is stable and clean, is agnostic to the webserver and doesn't rely on editing vendor files

I was under the impression both the xml and the routing file needed to be changed.
Read it wrong, thanks for clarifying.

**Comment by smnandre** [comment]
Question for people having trouble with this page... is the exception panel stylesheet working ? (it uses the same code as far as i know)

/_profiler/123456/exception.css

(with a profiler token instead of 123456)

**Comment by alexisLefebvre** [comment]
> Question for people having trouble with this page... is the exception panel stylesheet working ? (it uses the same code as far as i know)
> 
> /_profiler/123456/exception.css
> 
> (with a profiler token instead of 123456)

I'm not sure if that correspond to your question but the profiler and the exception pages are displayed without and with issues:

CSS is inlined on this page, it has no display issues:

![image](https://github.com/user-attachments/assets/f5131ee0-9fa7-47c8-baf8-5aee4747cf38)

CSS is broken on an exception page:

![image](https://github.com/user-attachments/assets/b74805d2-5540-48f5-b7c1-accebc9aa052)

**Comment by smnandre** [comment]
> CSS is inlined on this page, it has no display issues:

My PR had a bad impact on you people and i'm sorry.  Sincerely.

I'm trying to get all information i can, to find a good solution for everyone (for now: your suggestion to remove ".css" seems the best)


> I'm not sure if that correspond to your question but the pro

On your first screenshot, we see the profiler token was 970bae (the 6 digits in the URL)

Can you try accessing to  `https://localhost:8000/_profiler/970bae/exception.css`


Thanks

**Comment by alexisLefebvre** [comment]
> On your first screenshot, we see the profiler token was 970bae (the 6 digits in the URL)
> 
> Can you try accessing to `https://localhost:8000/_profiler/970bae/exception.css`
> 
> Thanks

Here is the result:

![image](https://github.com/user-attachments/assets/dbcbe2eb-19c4-4a07-8bf0-cd1efee2801e)

What page load this CSS file? I never saw an issue with `…/exception.css`.

**Comment by markuskuerten** [comment]
I encountered a similar problem after updating from 7.1 to 7.2.

On my site, the Symfony Toolbar initially appears without a stylesheet and only after a short delay does it display correctly with the stylesheet applied.

![grafik](https://github.com/user-attachments/assets/c6bc3cb9-5540-443c-a403-08769d2bb8a5)

For me, the only solution that worked was the workaround proposed by alexislefebvre "using the old version to override the new version."

https://github.com/symfony/symfony/issues/59045#issuecomment-2510191161

Here are some details about my environment:

Web server: nginx 1.24.0 (Ubuntu)
PHP version: php8.3-fpm
Proxy: None

I’d appreciate any guidance or a permanent fix for this issue.

**Comment by alexisLefebvre** [comment]
> I’d appreciate any guidance or a permanent fix for this issue.

This one is permanent: https://github.com/symfony/symfony/issues/59045#issuecomment-2523839740

Delete the cache with `rm -rf var/cache/*` if it doesn't work.

**Comment by markuskuerten** [comment]
> This one is permanent: [#59045 (comment)](https://github.com/symfony/symfony/issues/59045#issuecomment-2523839740)
> 
> Delete the cache with `rm -rf var/cache/*` if it doesn't work.

Thank you for your response and the suggestion! Unfortunately, clearing the cache with rm -rf var/cache/* doesn't resolve the issue in my case. I’ve already tried this approach several times, but the problem persists.

For now, the only solution that reliably works for me is the workaround I mentioned earlier (overriding the new version with the old one).

**Comment by smnandre** [comment]
There is a PR opened that fixes this problem. Just have to wait for it to be merged. 

https://github.com/symfony/symfony/pull/59229
